### PR TITLE
Define transpile optimization levels

### LINKED ISSUE #1943: move initializer optimizations to transpiler
### What is the expected enhancement?
The arbitrary state initializer is based on https://arxiv.org/abs/quant-ph/0406176v5, and it implements some of the optimizations that the paper talks about to reduce the depth. But the optimizations are ad-hoc, and they don't fully capture all optimization opportunities. We should make the synthesizer dumb, and relegate the optimizations to the transpiler.

For example, consider the circuit for initializing `[1/math.sqrt(2), 0, 0, 0, 0, 0, 0, 1/math.sqrt(2)]`.

The non-optimized/canonical synthesis is:
﻿image

The optimized version as implemented currently is:
image

Note that this manual optimization misses the fact that the last 4 CNOTs can be removed also.

If we just relegate all optimizations to the transpiler, then optimizing this circuit is fairly straightforward. It involves removing zero rotations, removing adjacent CNOTs, and commuting CNOTs that share a target but are on different qubits (these passes have to be done repetitively until converging to a fixed-point).

I'm removing these optimizations from the initializer as part of the refactoring in #1816. This issue should be fixed through the construction of a default pass manager that addresses the above.

### LINKED ISSUE #2057: Transpile unrolls to U gates even if gate is specified in basis_gates
### Information

- **Qiskit Terra version**: master
- **Python version**:
- **Operating system**:

### What is the current behavior?
```
q1 = QuantumRegister(1,'q1')
q2 = QuantumRegister(2,'q2')
q3 = QuantumRegister(3,'q3')
qc = QuantumCircuit(q1, q2, q3)
qc.h(q1[0])
qc.h(q2[1])
qc.h(q3[2])
layout = [4, 5, 6, 8, 9, 10]

cmap = [[1, 0], [1, 2], [2, 3], [4, 3], [4, 10], [5, 4], [5, 6], [5, 9],
 [6, 8], [7, 8], [9, 8], [9, 10], [11, 3], [11, 10], [11, 12], [12, 2], [13, 1],
 [13, 12]]

new_circ = transpile(qc, backend=None,
                     coupling_map=cmap,
                     basis_gates=['h'],
                     initial_layout=layout)
new_circ.data
```
```
[<qiskit.extensions.standard.u2.U2Gate at 0xd1c5cbcf8>,
 <qiskit.extensions.standard.u2.U2Gate at 0xd1c5cbe48>,
 <qiskit.extensions.standard.u2.U2Gate at 0xd1c5cbf60>]
```

### What is the expected behavior?
I would expect that defining a gate to be in the basis gates would not unroll to the U gate equivalent.

### LINKED ISSUE #2181: cx_direction pass should use U2 gates not H gates to avoid another unrolling
### What is the expected enhancement?
Currently the `cx_direction` pass uses Hadamard gates to reverse the direction of a CNOT.  However, this then requires another unrolling pass to the basis set of gates.  Instead, `cx_direction` should use the `U2(0,pi)` gate for the Hadamard (which of course are the same thing) so that the final unrolling is not needed.

### PR DESCRIPTION
**Pull request by ajavadia**
### Summary
Fixes #2057 
Fixes #2181 
Fixes #1943 

Expanding preset pass managers for improved optimization. These can be selected via:
`transpile(circuit, ..., optimization_level=level)`, where level can currently range from 0 to 3.

- Level 0 pass manager: no explicit optimization other than mapping to backend.
- Level 1 pass manager: light optimization by simple adjacent gate collapsing
- Level 2 pass manager: medium optimization by noise adaptive qubit mapping and
    gate cancellation using commutativity rules.
- Level 3 pass manager: heavy optimization by noise adaptive qubit mapping and
    gate cancellation using commutativity rules and unitary synthesis.

Higher levels in `transpile(optimization_level=...)` correspond to more aggressive optimization, but potentially slower.

I kept the default optimization flow (corresponding to `optimization_level=None`) as before to not interrupt the release.

### CODE REVIEW

File: qiskit/transpiler/transpile_circuit.py [hunk 1]
```diff
@@ -27,18 +32,31 @@ def transpile_circuit(circuit, transpile_config):
...
+        if level == 0:
+            pass_manager = level_0_pass_manager(transpile_config)
+        elif level == 1:
+            pass_manager = level_1_pass_manager(transpile_config)
+        elif level == 2:
+            pass_manager = level_2_pass_manager(transpile_config)
+        elif level == 3:
+            pass_manager = level_3_pass_manager(transpile_config)
+        else:
+            raise TranspilerError("optimization level can only be a non-negative integer")
```
Comment (maddy-tod) [inline qiskit/transpiler/transpile_circuit.py]: I think this error should say the range of allowable options, e.g. `optimization level can only be 0, 1, 2 or 3`,  as currently this wouldn't tell the user what was wrong if they set the optimization to `4`

File: qiskit/transpiler/passes/commutation_analysis.py [hunk 2]
```diff
@@ -73,7 +73,12 @@ def run(self, dag):
 
                 if current_gate not in current_comm_set[-1]:
                     prev_gate = current_comm_set[-1][-1]
-                    if _commute(current_gate, prev_gate):
+                    does_commute = False
```
Comment (atilag) [inline qiskit/transpiler/passes/commutation_analysis.py]: A simplification:
```python
try:
    if _commute(current_gate, prev_gate):
        current_comm_set[-1].append(current_gate)
except TranspilerError:
    pass
```

File: qiskit/transpiler/preset_passmanagers/level0.py [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+    _embed = [FullAncillaAllocation(coupling_map), EnlargeWithAncilla()]
+
+    # 3. Unroll to the basis
+    _unroll = Unroller(basis_gates)
+
+    # 4. Swap to fit the coupling map
+    _swap_check = CheckMap(coupling_map)
+
+    def _swap_condition(property_set):
+        return not property_set['is_swap_mapped']
```
Comment (atilag) [inline qiskit/transpiler/preset_passmanagers/level0.py]:
```suggestion
        return not property_set['is_swap_mapped']

```

File: qiskit/transpiler/preset_passmanagers/level0.py [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+    _swap = [BarrierBeforeFinalMeasurements(),
+             LegacySwap(coupling_map, trials=20, seed=seed_transpiler),
+             Decompose(SwapGate)]
+
+    # 5. Fix any bad CX directions
+    # _direction_check = CheckCXDirection(coupling_map)  # TODO
+    def _direction_condition(property_set):
+        return not property_set['is_direction_mapped']
```
Comment (atilag) [inline qiskit/transpiler/preset_passmanagers/level0.py]:
```suggestion
        return not property_set['is_direction_mapped']

```

File: qiskit/transpiler/preset_passmanagers/level0.py [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+    pm0 = PassManager()
+    if coupling_map:
+        pm0.append(_given_layout)
+        pm0.append(_choose_layout, condition=_choose_layout_condition)
+        pm0.append(_embed)
+    pm0.append(_unroll)
+    if coupling_map:
+        pm0.append(_swap_check)
+        pm0.append(_swap, condition=_swap_condition)
+        # pm0.append(_direction_check)
```
Comment (atilag) [inline qiskit/transpiler/preset_passmanagers/level0.py]:
```suggestion
```

File: qiskit/transpiler/preset_passmanagers/level1.py [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+    basis_gates = transpile_config.basis_gates
+    coupling_map = transpile_config.coupling_map
+    initial_layout = transpile_config.initial_layout
+    seed_transpiler = transpile_config.seed_transpiler
+
+    # 1. Use trivial layout if no layout given
+    _given_layout = SetLayout(initial_layout)
+
+    def _choose_layout_condition(property_set):
+        return not property_set['layout']
```
Comment (atilag) [inline qiskit/transpiler/preset_passmanagers/level1.py]:
```suggestion
        return not property_set['layout']

```

File: qiskit/transpiler/preset_passmanagers/level1.py [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+    _choose_layout = TrivialLayout(coupling_map)
+
+    # 2. Use a better layout on densely connected qubits, if circuit needs swaps
+    _layout_check = CheckMap(coupling_map)
+
+    def _improve_layout_condition(property_set):
+        return not property_set['is_swap_mapped']
```
Comment (atilag) [inline qiskit/transpiler/preset_passmanagers/level1.py]:
```suggestion
        return not property_set['is_swap_mapped']

```

File: qiskit/transpiler/preset_passmanagers/level1.py [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+    _embed = [FullAncillaAllocation(coupling_map), EnlargeWithAncilla()]
+
+    # 3. Unroll to the basis
+    _unroll = Unroller(basis_gates)
+
+    # 4. Swap to fit the coupling map
+    _swap_check = CheckMap(coupling_map)
+
+    def _swap_condition(property_set):
+        return not property_set['is_swap_mapped']
```
Comment (atilag) [inline qiskit/transpiler/preset_passmanagers/level1.py]:
```suggestion
        return not property_set['is_swap_mapped']

```

File: qiskit/transpiler/preset_passmanagers/level1.py [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+    _swap = [BarrierBeforeFinalMeasurements(),
+             LegacySwap(coupling_map, trials=20, seed=seed_transpiler),
+             Decompose(SwapGate)]
+
+    # 5. Fix any bad CX directions
+    # _direction_check = CheckCXDirection(coupling_map)  # TODO
+    def _direction_condition(property_set):
+        return not property_set['is_direction_mapped']
```
Comment (atilag) [inline qiskit/transpiler/preset_passmanagers/level1.py]:
```suggestion
        return not property_set['is_direction_mapped']

```

File: qiskit/transpiler/preset_passmanagers/level1.py [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+    _direction = [CXDirection(coupling_map)]
+
+    # 6. Remove zero-state reset
+    _reset = RemoveResetInZeroState()
+
+    # 7. Merge 1q rotations and cancel CNOT gates iteratively until no more change in depth
+    _depth_check = [Depth(), FixedPoint('depth')]
+
+    def _opt_control(property_set):
+        return not property_set['depth_fixed_point']
```
Comment (atilag) [inline qiskit/transpiler/preset_passmanagers/level1.py]:
```suggestion
        return not property_set['depth_fixed_point']
        
```

File: qiskit/transpiler/preset_passmanagers/level1.py [hunk 11]
```diff
@@ -0,0 +1,129 @@
...
+        pm1.append(_given_layout)
+        pm1.append(_choose_layout, condition=_choose_layout_condition)
+        pm1.append(_layout_check)
+        pm1.append(_improve_layout, condition=_improve_layout_condition)
+        pm1.append(_embed)
+    pm1.append(_unroll)
+    if coupling_map:
+        pm1.append(_swap_check)
+        pm1.append(_swap, condition=_swap_condition)
+        # pm1.append(_direction_check)
```
Comment (atilag) [inline qiskit/transpiler/preset_passmanagers/level1.py]:
```suggestion
```

File: qiskit/transpiler/preset_passmanagers/level2.py [hunk 12]
```diff
@@ -0,0 +1,127 @@
...
+    coupling_map = transpile_config.coupling_map
+    initial_layout = transpile_config.initial_layout
+    seed_transpiler = transpile_config.seed_transpiler
+    backend_properties = transpile_config.backend_properties
+
+    # 1. Layout on good qubits if calibration info available, otherwise on dense links
+    _given_layout = SetLayout(initial_layout)
+
+    def _choose_layout_condition(property_set):
+        return not property_set['layout']
```
Comment (atilag) [inline qiskit/transpiler/preset_passmanagers/level2.py]:
```suggestion
        return not property_set['layout']

```

File: qiskit/transpiler/preset_passmanagers/level2.py [hunk 13]
```diff
@@ -0,0 +1,127 @@
...
+        _choose_layout = NoiseAdaptiveLayout(backend_properties)
+
+    # 2. Extend dag/layout with ancillas using the full coupling map
+    _embed = [FullAncillaAllocation(coupling_map), EnlargeWithAncilla()]
+
+    # 3. Unroll to 1q or 2q gates, swap to fit the coupling map
+    _swap_check = CheckMap(coupling_map)
+
+    def _swap_condition(property_set):
+        return not property_set['is_swap_mapped']
```
Comment (atilag) [inline qiskit/transpiler/preset_passmanagers/level2.py]:
```suggestion
         return not property_set['is_swap_mapped']

```

File: qiskit/transpiler/preset_passmanagers/level2.py [hunk 14]
```diff
@@ -0,0 +1,127 @@
...
+             LegacySwap(coupling_map),
+             Decompose(SwapGate)]
+
+    # 4. Unroll to the basis
+    _unroll = Unroller(basis_gates)
+
+    # 5. Fix any bad CX directions
+    # _direction_check = CheckCXDirection(coupling_map)  # TODO
+    def _direction_condition(property_set):
+        return not property_set['is_direction_mapped']
```
Comment (atilag) [inline qiskit/transpiler/preset_passmanagers/level2.py]:
```suggestion
        return not property_set['is_direction_mapped']
        
 ```

File: qiskit/transpiler/preset_passmanagers/level2.py [hunk 15]
```diff
@@ -0,0 +1,127 @@
...
+    _direction = [CXDirection(coupling_map)]
+
+    # 6. Remove zero-state reset
+    _reset = RemoveResetInZeroState()
+
+    # 7. 1q rotation merge and commutative cancellation iteratively until no more change in depth
+    _depth_check = [Depth(), FixedPoint('depth')]
+
+    def _opt_control(property_set):
+        return not property_set['depth_fixed_point']
```
Comment (atilag) [inline qiskit/transpiler/preset_passmanagers/level2.py]:
```suggestion
         return not property_set['depth_fixed_point']

```

File: qiskit/transpiler/preset_passmanagers/level2.py [hunk 16]
```diff
@@ -0,0 +1,127 @@
...
+    pm2 = PassManager()
+    if coupling_map:
+        pm2.append(_given_layout)
+        pm2.append(_choose_layout, condition=_choose_layout_condition)
+        pm2.append(_embed)
+    pm2.append(_unroll)
+    if coupling_map:
+        pm2.append(_swap_check)
+        pm2.append(_swap, condition=_swap_condition)
+        # pm2.append(_direction_check)
```
Comment (atilag) [inline qiskit/transpiler/preset_passmanagers/level2.py]:
```suggestion
```

File: qiskit/transpiler/preset_passmanagers/level3.py [hunk 17]
```diff
@@ -0,0 +1,120 @@
...
+    coupling_map = transpile_config.coupling_map
+    initial_layout = transpile_config.initial_layout
+    seed_transpiler = transpile_config.seed_transpiler
+    backend_properties = transpile_config.backend_properties
+
+    # 1. Layout on good qubits if calibration info available, otherwise on dense links
+    _given_layout = SetLayout(initial_layout)
+
+    def _choose_layout_condition(property_set):
+        return not property_set['layout']
```
Comment (atilag) [inline qiskit/transpiler/preset_passmanagers/level3.py]:
```suggestion
        return not property_set['layout']

```

File: qiskit/transpiler/preset_passmanagers/level3.py [hunk 18]
```diff
@@ -0,0 +1,120 @@
...
+        _choose_layout = NoiseAdaptiveLayout(backend_properties)
+
+    # 2. Extend dag/layout with ancillas using the full coupling map
+    _embed = [FullAncillaAllocation(coupling_map), EnlargeWithAncilla()]
+
+    # 3. Unroll to 1q or 2q gates, swap to fit the coupling map
+    _swap_check = CheckMap(coupling_map)
+
+    def _swap_condition(property_set):
+        return not property_set['is_swap_mapped']
```
Comment (atilag) [inline qiskit/transpiler/preset_passmanagers/level3.py]:
```suggestion
        return not property_set['is_swap_mapped']

```

File: qiskit/transpiler/preset_passmanagers/level3.py [hunk 19]
```diff
@@ -0,0 +1,120 @@
...
+             LegacySwap(coupling_map)]
+
+    # 4. Unroll to the basis
+    _unroll = Unroller(basis_gates)
+
+    # 5. 1q rotation merge and commutative cancellation iteratively until no more change in depth
+    _depth_check = [Depth(), FixedPoint('depth')]
+
+    def _opt_control(property_set):
+        return not property_set['depth_fixed_point']
```
Comment (atilag) [inline qiskit/transpiler/preset_passmanagers/level3.py]:
```suggestion
        return not property_set['depth_fixed_point']

```

File: qiskit/transpiler/passes/commutation_analysis.py (hunk 2 above)
Comment (ajavadia) [reply qiskit/transpiler/passes/commutation_analysis.py]: not sure if this is the same as the `else` falls outside the try/except.

Comment (atilag) [reply qiskit/transpiler/passes/commutation_analysis.py]: You are right, the blank line confused me.. let's leave it as it is.

**Pull request merged**